kani-compiler/src/kani_middle/coercion.rs

```diff
@@ -0,0 +1,232 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! This module contains methods that help us process coercions.
+//! There are many types of coercions in rust, they are described in the
+//! [RFC 401 Coercions](https://rust-lang.github.io/rfcs/0401-coercions.html).
+//!
+//! The more complicated coercions are DST Coercions (aka Unsized Coercions). These coercions
+//! allow rust to create references to dynamically sized types (such as Traits and Slices) by
+//! casting concrete sized types. Unsized coercions can also be used to cast unsized to unsized
+//! types. These casts work not only on the top of references, but it also handle
+//! references inside structures, allowing the unsized coercions of smart pointers. This last
+//! piece was defined by
+//! [RFC 982 DST Coercion](https://rust-lang.github.io/rfcs/0982-dst-coercion.html).
+
+use rustc_hir::lang_items::LangItem;
+use rustc_middle::traits::{ImplSource, ImplSourceUserDefinedData};
+use rustc_middle::ty::adjustment::CustomCoerceUnsized;
+use rustc_middle::ty::TypeAndMut;
+use rustc_middle::ty::{self, ParamEnv, TraitRef, Ty, TyCtxt};
+use rustc_span::symbol::Symbol;
+use tracing::trace;
+
+/// Extract the pair (`T`, `U`) for a unsized coercion where type `T` implements `Unsize<U>`.
+/// I.e., `U` is either a trait or a slice.
+/// For more details, please refer to:
+/// <https://doc.rust-lang.org/reference/type-coercions.html#unsized-coercions>
+///
+/// This is used to determine the vtable implementation that must be tracked by the fat pointer.
+///
+/// For example, if `&u8` is being converted to `&dyn Debug`, this method would return:
+/// `(u8, dyn Debug)`.
```

Comment:
I wonder if we could rephrase this opening to start with the concrete example instead of only coming to it later.

> Given an unsized coercion (e.g. from `&u8` to `&dyn Debug`), extract the pair of coresponding base types `T, U` (e.g. `u8`, `dyn Debug`), where the source base type `T` must implement `Unsize<U>` and `U` is either a trait or slice.

Reply:
sure